File: lang/literal.rs
trait IsNan {{
    fn is_nan(&self) -> bool {{
        false
    }}
}}

impl IsNan for i64 {{
}}

pub fn main() {{
    let x: {type} = {value};
    let expected: {type} = {expected};
    if expected.is_nan() {{
        assert_eq!(x.is_nan(), expected.is_nan());
    }} else {{
        assert_eq!(x, expected);
    }}
}}

File: lang/literal.cpp
#include <cassert>
#include <cmath>
#include <stdint.h> // use for pre-C++11 compatability

typedef int64_t i64;
typedef uint64_t u64;
typedef double f64;

int main() {{
    {type} actual = {value};
    {type} expected = {expected};
    if (expected != expected) {{
        assert(actual != actual);
    }} else {{
        assert(actual == expected);
    }}
    return 0;
}}
